k-distribution/src/main/scripts/lib/pyk/pyk/kcfg.py

```diff
@@ -1,59 +1,105 @@
 import json
-import threading
-from typing import Final
+from abc import ABC
+from dataclasses import dataclass
+from functools import cached_property, reduce
+from itertools import chain
+from threading import RLock
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+)
 
 from graphviz import Digraph
 
-from .cli_utils import fatal, notif
-from .kast import KClaim, KFlatModule, KImport, KInner, KRule
+from .cterm import CTerm
+from .kast import TRUE, KInner, KRuleLike
 from .kastManip import (
+    Subst,
     buildRule,
-    matchWithConstraint,
-    minimizeSubst,
     mlAnd,
     simplifyBool,
-    substitute,
-    substToMap,
     unsafeMlPredToBool,
 )
-from .utils import compare_short_hashes, hash_str, shorten_hashes
+from .ktool import KPrint
+from .utils import compare_short_hashes, shorten_hashes
 
 
 class KCFG:
-    _FIELDS: Final = (
-        'states',
-        'graph',
-        'abstractions',
-        'loop',
-        'split',
-        'init',
-        'target',
-        'terminal',
-        'stuck',
-        'subsumptions',
-        'frontier',
-    )
-
-    _NODE_ATTRS: Final = ('init', 'target', 'stuck', 'terminal', 'frontier', 'loop', 'split')
+
+    @dataclass(frozen=True)
+    class Node:
+        term: CTerm
+
+        def __init__(self, term: CTerm):
+            object.__setattr__(self, 'term', term)
+
+        @property
+        def id(self) -> str:
+            return self.term.hash
+
+        def to_dict(self) -> Dict[str, Any]:
+            return {'id': self.id, 'term': self.term.term.to_dict()}
+
+    class EdgeLike(ABC):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
+
+    @dataclass(frozen=True)
+    class Edge(EdgeLike):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
+        condition: KInner = TRUE
+        depth: int = 1
+
+        def to_dict(self) -> Dict[str, Any]:
+            return {'source': self.source.id, 'target': self.target.id, 'condition': self.condition.to_dict(), 'depth': self.depth}
+
+        def to_rule(self, claim=False, priority=50) -> KRuleLike:
+            sentence_id = f'BASIC-BLOCK-{self.source.id}-TO-{self.target.id}'
+            init_term = mlAnd([self.source.term, self.condition])
+            final_term = self.target.term
+            return buildRule(sentence_id, init_term, final_term, claim=claim, priority=priority)
+
+    @dataclass(frozen=True)
+    class Cover(EdgeLike):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
+
+        @cached_property
+        def subst(self) -> Optional[Subst]:
+            match_res = self.source.term.match(self.target.term)
+            if not match_res:
+                return None
```

Comment:
Every cover should have a substitution. If `match_res is None`, it sohuld probably be a fatal error I think?

Reply:
Fixed.